Define custom shapes

SjurdurS opened the pull request:
This pull request adds a new Custom Shape Tool, called "Shape", to the Level Editor that allows you to quickly place predefined custom shapes into the current level.
Furthermore, this pull request also adds a new right click context menu option called "Save as shape..." which creates a new custom shape based on the current selection.

Custom shapes are saved as regular Elma levels (*.lev) on disk, and are grouped by sub-folders in the root folder for custom shapes. The root folder for custom shapes is called `sle_shapes` and will be created in the same directory as Elmanager.exe the first time a custom shape is saved.

The custom shape collection is just a folder structure with files, and therefore easily shareable / importable / exportable.

The following components can be added to custom shapes:
 - Polygons (always full polygons, if any vertex is selected on a polygon, the whole polygon is serialized).
 - Grass polygons
 - Objects (Apples, Killers, Flowers)
 - Pictures
 - Textures

The new Custom Shape Tool supports the following actions:
 - Right click to open the Shape Selection dialog
 - Anchoring using num 1-5 like Picture Tool.
 - Transformation
   - Mirroring - Vertical, Horizontal or Both (toggle with num 6)
   - Rotation - num 7 to rotate left, num 9 to rotate right, num 8 to reset rotation
   - Scaling - Resize using (+ / - ) on the keyboard
   - Reset anchoring and transformations: num 0
   - The transformation state is remembered between shapes, so if you rotate one shape, and select a different one, the same settings apply to the new shape as well.

# Images
### New Shape button, and Shape Selection dialog
![Elmanager_6LHlIvoMYi](https://github.com/user-attachments/assets/1e297d05-39b6-4046-8d23-6adb9c14d818)


### Shapes are grouped by folders.
![Elmanager_ckvefUHXjf](https://github.com/user-attachments/assets/85795a18-a164-4269-a4cd-2523d8627456)



# Example shapes for quick testing
Here is an example collection of shapes you can use for trying it out.
Place the following `sle_shapes` folder in the same folder as the Elmanager.exe
[sle_shapes.zip](https://github.com/user-attachments/files/19263475/sle_shapes.zip)

SjurdurS committed:
Add support for creating and loading custom shapes

SjurdurS committed:
Fix crash when gallery directory doesn't exist

Smibu commented:
Wow nice! I'll take a look today.

Smibu reviewed (commented):
I added a few comments in the code, but at the same time I'd like to suggest a hopefully simpler implementation:

* Instead of storing the shapes as JSON files, could they be just normal lev files? This way they would be easily editable, and we wouldn't need custom serialization logic for them.
* Instead of storing a cached image of the shape, could we implement a `LevelControl` that inherits from `GLControl` that is able to directly render a given lev? So a CustomShapeControl would have a reference to a LevelControl instead of a PictureBox. I imagine this might not be a huge task because all the pieces are there already; we just need to glue them together. We could share the main window GL context. This maybe requires updating GLControl to the [official version](https://github.com/opentk/GLControl) because it seems the current (vendored in repo) version might not support GL context sharing.

SjurdurS commented on Elmanager/LevelEditor/LevelEditorForm.cs:
```diff
@@ -2953,4 +2977,152 @@ private void fixSelfIntersectionsMenuItem_Click(object sender, EventArgs e)
+
+        var (center, _, _) = GeometryUtils.CalculateBoundingBox(clonedPolygons, clonedObjects, clonedGraphicElements);
+
+        // Normalize positions
```
I just felt it would be easier, and more clean that shapes are centered around origin than arbitrary positions.

Also, you need to do this step during translation and rotation, so doing it once before saving, instead of each time the shape is in use reduces the amount of calculations you need to do, unless I'm mistaken?

Smibu replied:
Yes that makes sense, but I think the normalization should probably be done when *loading* the shapes, rather than on save. Especially now that the shapes are editable as levs, we cannot assume the coordinates are normalized in the shape (lev) file.

SjurdurS replied:
You are right. I'll look into changing it to on loading instead.

SjurdurS commented:
Thank you looking at my changes, and thanks for the great feedback! :)

Using the LEV format makes sense, especially if they can be rendered directly in the selection dialog, but I have no experience with OpenGL which is why I didn't go down that road.

I think it would make sense to do this in two steps.
1. First replace the JSON serialization with the LEV format, but keep the images for now.
2. Then afterwards try to render the levels in a LevelControl.

SjurdurS committed:
Use generic Enum.GetValues for ShapeMirrorOption

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Shapes are now serialized using the LEV format

I haven't removed the support for deserialization of JSON  shapes to help with converting between the old JSON and the new LEV format.

SjurdurS committed:
Add required start object to LEV shape.

SjurdurS commented:
I did a quick change to serialize to and from LEV instead of JSON. Using LEV format works well.
I will still need to clean up and remove all the unnecessary code related to the JSON serialization.

I noticed that LEV objects require a start object (I got a random crash during conversion from JSON to LEV). It isn't a problem adding a start object as it is easy to filter them on load, but am I missing any other requirements for a Level to be "valid"?

Here is an updated example gallery should you want to test it.
[GalleryWithLevFormat.zip](https://github.com/user-attachments/files/18547903/GalleryWithLevFormat.zip)

Smibu commented:
I think just the start object and 1 (ground) polygon is required for validity.

By two steps, do you mean two PRs? I would like just this one PR where we do both steps. You can regardless send some intermediate version for people to test in discord if you want to gather early feedback.

SjurdurS commented:
> I think just the start object and 1 (ground) polygon is required for validity.
Then the easiest solution is to limit custom shape creation to when the user has selected at least 1 polygon.
With my JSON implementation you could create shapes out of anything, pictures only, objects only etc. But this is not necessary.
Otherwise you could add some kind of identifiable placeholder polygon that you remove on load, but for now I'd rather just stick with the minimum.

 
> By two steps, do you mean two PRs? I would like just this one PR where we do both steps. You can regardless send some intermediate version for people to test in discord if you want to gather early feedback.
No, I just mean refactoring-wise. Refactor first to use LEV, and get rid of all of the unnecessary JSON serialization related code, and only then try to implement the Level rendering. Since I have 0 experience with GL, I don't know if I will be able to implement the last part :)

SjurdurS committed:
Only enable Create Custom Shape menu item when at least 1 ground polygon is selected

Smibu commented:
Ah ye ok, that sounds good 👍

If the GL thing turns out difficult, I can take a look at it sometime. The new control might be useful in other contexts also.

SjurdurS committed:
Prevent creating custom shape when only grass polygons are selected

SjurdurS committed:
Refactor to remove unnecessary JSON deserialization

SjurdurS committed:
Normalize positions on load instead of before saving

SjurdurS committed:
Refactor to not use DataSource on ComboBox

SjurdurS commented:
> Ah ye ok, that sounds good 👍
> 
> If the GL thing turns out difficult, I can take a look at it sometime. The new control might be useful in other contexts also.

I haven’t started looking at it yet, but I have a question regarding how you see the LevelControl used. If you have a grid like the one I used in the shape selection dialog. Would the grid just consist of individual LevelControls, where each LevelControl has its own shared GL context? How would performance be in such a case?

Smibu commented:
The grid would consist of CustomShapeControls (as it already is now?) and a CustomShapeControl would have a reference to a LevelControl (instead of a PictureBox).

So you'd pass the LevelEditorForm's GLControl to ShapeGalleryForm so that it can be further passed to each CustomShapeControl and LevelControl as the shared context.

I wouldn't expect performance problems; GL rendering should be pretty fast and hopefully the GLControl's paint method is only invoked when the control is actually visible (and not e.g. outside of a scroll area etc).

SjurdurS committed:
Trying to get LevelControl to work ..

SjurdurS commented:
I've played around with making a minimal example of a custom GLControl, and trying to just draw a triangle in a form and I can't even get anything drawn on screen.

I tried changing to OpenTK.GLControl throughout, as it has SharedContext defined, and the Editor itself draws as usual. But I just get InvalidOperation errors in the test form/dialog.

Really at a loss for what to do here .. any pointers? :D

SjurdurS committed:
Test

Smibu commented:
I need to see some code, hard to say otherwise. Maybe push some wip branch to your fork?

SjurdurS committed:
WIP - testing to get minimal working example of a shared context

SjurdurS commented:
> I need to see some code, hard to say otherwise. Maybe push some wip branch to your fork?

See a very sloppy test implementation at: 
https://github.com/Smibu/elmanager/commit/2bcb9e92e0f59627819f15fe514e77696c94bcd0

The dialog is loaded when you trigger Create Custom Shape, but you can move it elsewhere if necessary.

I just wanted to create a minimal working example, where I'm using the shared context to do a basic drawing.
I managed to get it to draw a red background on the dialog, in the "LevelControl", but in the render loop I get nonstop InvalidOperation.

I had to install a different package "OpenTK.GLControl" to get the SharedContext property in GLControl.
Many examples online also refer to a "GraphicsMode" enum and other classes not available in the packages you are using.

Smibu committed:
LevelControl fixes

Smibu commented:
Thanks :) it works with a few changes: https://github.com/Smibu/elmanager/commit/d71f140bfbd3c2f63f84232cbe005608f74ef36c

![image](https://github.com/user-attachments/assets/71eb7d1a-1f7f-47fd-8128-ec8d7bc58565)

It was just missing the compat profile:

```csharp
Profile = ContextProfile.Compatability;
```

(Works also with the timer thing, but probably OnPaint is enough, as shapes are just static drawings)

SjurdurS commented:
Great! I’ll try to get level rendering to work when I have time later this week. The timer thingy was just me being desperate to get something calling as I had problems with the OnPaint method initially.

SjurdurS committed:
wip level rendering

SjurdurS committed:
level renders in control now

SjurdurS committed:
wip - levels can now be rendered in the shape gallery, but performance is horrendous

SjurdurS committed:
Slightly faster shape LevelControl initialization, but now shape backgrounds are red

SjurdurS committed:
Cloning OpenGlLgr doesn't work as is

SjurdurS committed:
Update comment

SjurdurS commented:
So I managed to render the shapes in the Shape Gallery using the Shared Context, but the loading of multiple shapes in the Shape Gallery is really slow (several seconds to load and render 10-20 shapes).

Currently one of the bottlenecks is the call `_renderer.UpdateSettings(_level, _renderingSettings);` I use on LevelControl load.

Any suggestions on how to share the resources used in an efficient way?

```
  private void LevelControl_Load(object? sender, EventArgs e)
  {
      // Initialization code for OpenGL
      _renderer = new ElmaRenderer(this, _renderingSettings);
      //_renderer.OpenGlLgr = _originalElmaRenderer.OpenGlLgr; // Shapes are not rendered correctly with this line
      var r = _renderer.UpdateSettings(_level, _renderingSettings); // << this line is slow, since it initializes OpenGlLgr from scratch every time a LevelControl is initialized
      _renderer.InitializeLevel(_level, _renderingSettings);
      _level.UpdateBounds();
      _zoomController.ZoomFill(_renderingSettings);

  }
```

FYI: I continued on the same test branch as before. If I get everything working I can clean up the code and move it to the pull request branch.

I'm aware I need to adjust the rendering settings, hide the kuski etc, but that is something I will focus on after I get it working properly.


Screenshots:

![Elmanager_janc9GqvYx](https://github.com/user-attachments/assets/8a7251f1-28c5-43a4-8919-4f788dd761f8)


![Elmanager_7JH8hne0Di](https://github.com/user-attachments/assets/af17b458-6aff-4e45-a623-a44eac93eea3)

SjurdurS committed:
More testing. Still slow in debug

SjurdurS committed:
Small adjustments

SjurdurS committed:
Fix size of LevelControl

SjurdurS committed:
Adjust line width

SjurdurS committed:
Refactor first time paint handling and zoom controller usage

SjurdurS commented:
Initializing many GLControls seems very slow. Perhaps the best solution is to use a single GLControl to render all shapes in some kind of grid like manner. However, this is easier said than done :S

Smibu commented:
Nice progress!

GLControlSettings has AutoLoadBindings property that is true by default. Try setting it to false for LevelControls? Not sure if it helps, but it seems worth trying.

SjurdurS commented:
I can give it a try.

if I can’t make it perform well enough then I would like to try something else.

Instead of initializing arbitrarily many GLControls, it might be simpler to just generate image snapshots when loading the shapes, and displaying the snapshots instead of using live gl rendering.
So the shape gallery Form would use 1 instance of LevelControl to render each shape one by one to  images in memory. Then add those images to the grid view.

If the user wants to view the shapes with a different LGR or other settings you would just re-render each snapshot.

SjurdurS committed:
Try disabling AutoLoadBindins

SjurdurS commented:
> Nice progress!
> 
> GLControlSettings has AutoLoadBindings property that is true by default. Try setting it to false for LevelControls? Not sure if it helps, but it seems worth trying.

Changing AutoLoadBindings makes no measurable difference.

For 5-10 LevelControls the performance is ok, but as the number of shapes goes up, to say 40, it takes several seconds to load. So it doesn’t scale well as it is now.

SjurdurS committed:
Work towards implementing a virtual mode panel for the shape gallery

SjurdurS committed:
Work towards implementing a virtual mode panel for the shape gallery

SjurdurS committed:
Work towards implementing a virtual mode panel for the shape gallery

SjurdurS committed:
Reuse UI controls when switching shape groups for increased performance

SjurdurS committed:
Add custom scroll bar

SjurdurS committed:
Refactor to use a virtual mode flow panel

SjurdurS committed:
Add mouse wheel scroll to flow panel

SjurdurS committed:
Disable resizing of shape gallery form to simplify handling of the virtual mode flow panel

SjurdurS committed:
Remove unused classes

SjurdurS committed:
Fix compiler warnings

SjurdurS committed:
Add function to disable level control rendering on demand

SjurdurS committed:
Update comment

SjurdurS commented:
In the following branch I've tried to refactor the flow layout panel to a virtual mode panel or whatever it is called, where I reuse the UI controls to render the shapes and do a manual handling of scrolling . The performance is much better, and changing between groups is almost instantaneous.

The startup of the Shape Gallery Form is a bit slow (especially in Debug mode and if you don't use the commented lines below), but it usable now imo.

        private void LevelControl_Load(object? sender, EventArgs e)
        {
            // Initialization code for OpenGL
            _renderer = new ElmaRenderer(this, _renderingSettings);
            _renderer.OpenGlLgr = _originalElmaRenderer.OpenGlLgr; // Slightly faster with these lines. <<
            _renderer._lgrCache = _originalElmaRenderer._lgrCache; // Slightly faster with these lines. <<
        }

There is however still some kind of memory leak, I think it's related to ElmaRenderer / OpenGlLgr (Textures).
I haven't looked much into this, but I did a quick test of calling dispose on the above objects, but I just got crashes.

See the branch:
https://github.com/SjurdurS/elmanager/tree/refactor_to_virtual_mode_panel

Smibu commented:
Very nice, I think this is the right direction :)

Layout is a bit broken for me - it looks like this when I've scrolled all the way down. Could be related to different screen DPI because it has 7 controls per row now. Also labels seem to clip off partially.

![image](https://github.com/user-attachments/assets/a907ed3d-b377-4153-bbfa-19e52bea39a1)

One idea for memory leak: try setting DisableFrameBuffer = true for RenderingSettings in ShapeGalleryForm. That buffer is only used for snapshots which aren't needed here.

SjurdurS commented:
> One idea for memory leak: try setting DisableFrameBuffer = true for RenderingSettings in ShapeGalleryForm. That buffer is only used for snapshots which aren't needed here.

Looks like disabling frame buffer solves the memory leak. 

> Layout is a bit broken for me - it looks like this when I've scrolled all the way down. Could be related to different screen DPI because it has 7 controls per row now. Also labels seem to clip off partially.

I haven't tested high DPI at all, and also I don't know how .net handles DPI scaling compared to MFC.
The current layout of the Shape form is a hard coded in pixels to fit 4x3 shapes. The virtual implementation of the grid is also hard coded to be exactly 4x3 :) I'll see what I can do about it.

SjurdurS committed:
Fix memory leak

SjurdurS committed:
Change auto scaling mode to Dpi

SjurdurS commented:
I can't reproduce the scaling issue. I don't have a 4K monitor, but I tried to change display scaling to 150%-225% and it still looked fine on my computer (Win 10).

Can you check if the scaling issue is fixed with the following change?
https://github.com/Smibu/elmanager/commit/3e82e9aceb794adcd0b87eccaa45897dfbf6a0ba

Smibu commented:
> Looks like disabling frame buffer solves the memory leak.

👍

> Can you check if the scaling issue is fixed with the following change?

No, it looks almost the same still - now it has 8 shapes per row.

I don't have ideas for this offhand. If nothing else, maybe it can just be some if-elses based on DPI. I can investigate it sometime, it's easier for me to try stuff when it reproes.

SjurdurS committed:
Improve DPI scaling of Shape Gallery Form when "fix blurry apps" setting in Windows is enabled

SjurdurS commented:
> > Can you check if the scaling issue is fixed with the following change?
> 
> No, it looks almost the same still - now it has 8 shapes per row.
> 
> I don't have ideas for this offhand. If nothing else, maybe it can just be some if-elses based on DPI. I can investigate it sometime, it's easier for me to try stuff when it reproes.

I managed to reproduce the issue by enabling _"Let Windows try to fix apps so they're not blurry_" in **Display Settings** > **Advanced scaling settings**.

Scaling should be better now in the latest commit I made: https://github.com/Smibu/elmanager/commit/7e29bbd7fd30058559f6aa482ec30b60895cd881

Here is a screenshot at 200% scaling.
<img width="1276" alt="Elmanager_oVWUEzSgYy" src="https://github.com/user-attachments/assets/67a2bf9e-f6cd-4dfe-9f85-f9618a2feb93" />

Smibu commented:
Nice! Didn't know about that Windows setting. Works now for me too.

I'd suggest a couple of changes to the gallery window:
* It could be a little smaller (I mean smaller LevelControls, for example 75% of current width/height). Or ideally, resizable window (but still static number of LevelControls, 4x3).
* Remove maximize and minimize buttons (same as other tool windows are).
* Remove the factor/rotation/mirror/reset options because you can adjust them while placing, and you can't see preview while window is open anyway, so it's hard to guess good settings at first try.
* Maybe double-click could mean "Select + OK" as a quick shortcut.

SjurdurS commented:
Good suggestions. I’d like to minimize the scope until everything is properly implemented, so removing bloat is a welcome idea :)

However rendering the rotation/mirroring in the LevelControl would be possible now if one really wanted to, but better to keep it simple for now :)

I also noticed the scroll bar doesn’t scale correctly at higher DPI.

Do you have any suggestions / opinions on what this while thing should be called, both in the code and in the UI?

SjurdurS committed:
Remove Maximize and Minimize buttons from caption bar

SjurdurS committed:
Hide icon from title bar

SjurdurS committed:
Remove the factor/rotation/mirror/reset options from Shape Gallery Form

Smibu commented:
> Do you have any suggestions / opinions on what this while thing should be called, both in the code and in the UI?

Shape is probably just fine. So tool name can be "Shape" as it is now and window title could be simply "Select shape". I'd remove the "gallery" word everywhere. And the root dir name could be "sle_shapes". Some people may have elmanager.exe in eol dir, so it's probably good to spell out that the dir is related to SLE.

SjurdurS committed:
Added option to double click a shape to quickly select and close the Shape Gallery Form dialog

SjurdurS committed:
Remove unused parameters

SjurdurS committed:
Remove no longer needed code for taking bitmap snap shot of shape

SjurdurS committed:
Remove no longer needed code to zoom to selection

This was used for taking shape snap shots of the selection

SjurdurS committed:
Remove no longer needed save and take snapshots functions

These were used for taking bitmap snap shots for shapes

SjurdurS committed:
Remove unused using statements

SjurdurS committed:
Code cleanup

SjurdurS committed:
Normalize positions by using level bounds after having removed start from the loaded shape

SjurdurS committed:
Refactor to not modify the level in LevelControl before rendering

SjurdurS committed:
Rename ShapeGallery names to Shapes

SjurdurS committed:
Rename Shape Gallery Form to Shape Selection Form

SjurdurS committed:
Fix incorrect references to images

SjurdurS committed:
Refactor variable name from gallery to shapes

SjurdurS committed:
Change default shapes directory path from Gallery to sle_shapes

SjurdurS committed:
Remove commented out using statement

SjurdurS committed:
Programatically set shape selection form size

SjurdurS committed:
Adjust layout in Shape Selection Form

SjurdurS committed:
Remove commented out code

SjurdurS committed:
Adjust layout in Shape Selection Form

SjurdurS committed:
Enable scroll bar only when there is something to scroll

SjurdurS committed:
Refactor to use TableLayoutPanel instead of FlowLayoutPanel

SjurdurS committed:
Allow for resizing of shape seleciton form

LevelControls auto resize as the form resizes

SjurdurS committed:
Debounce resize events to reduce rendering calls

Continously re-rendering on resize is rather slow.

SjurdurS committed:
Remove references to the word Gallery

SjurdurS commented:
> It could be a little smaller (I mean smaller LevelControls, for example 75% of current width/height). Or ideally, resizable window (but still static number of LevelControls, 4x3).

Default size is smaller now, and resizing is possible.
Performance of re-rendering was a bit slow on my machine, so I implemented debouncing of resize events to try to reduce the amounts of render calls. It seems to be a bit better now, at the cost of a bit "laggy" UI update.

> Remove maximize and minimize buttons (same as other tool windows are).

Done

> Remove the factor/rotation/mirror/reset options because you can adjust them while placing, and you can't see preview while window is open anyway, so it's hard to guess good settings at first try.

Done

> Maybe double-click could mean "Select + OK" as a quick shortcut.

Done

> Shape is probably just fine. So tool name can be "Shape" as it is now and window title could be simply "Select shape".

I renamed the form to Shape Selection Form, and the title is now "Shape Selection". I also refactored the names in the code.

> I'd remove the "gallery" word everywhere. And the root dir name could be "sle_shapes".

I removed all references to the word "gallery", and changed the default root directory to "sle_shapes".

Here's a short demo of the latest changes:
https://github.com/user-attachments/assets/cfb1d59d-404f-441c-932d-e9dbdb2292a6

Smibu commented:
Nice, some quick feedback about the window:

Default window size seems too small for me:

![image](https://github.com/user-attachments/assets/92b92cca-0439-4b5e-a32c-4dcd68527c11)

(Also it doesn't remember the new size if I resize and close it)

There is also some bug with zoom filling, I easily get something like this:

![image](https://github.com/user-attachments/assets/20d4cb2e-29ea-4f28-b343-719158e54d7d)

SjurdurS commented:
Remembering size I didn’t touch yet since I couldn’t decide if the state should be per session or persistent.

Per session is probably good enough?

Also didn’t get around to testing so much yet :) especially didn’t test any DPI scaling after all the refactoring.

Smibu commented:
Yea, no hurry with testing.

Per session is ok. Good default size is more important.

SjurdurS committed:
Fix sizing issue of Custom Shape Control

SjurdurS committed:
Remove the use of margins and paddings on the Level Control itself

Margins and Location are inconsistently scaled at higher DPI so the location of the control and margins are not matching at certain DPI values.

SjurdurS committed:
Remove no longer needed size calculation of Shape Selection Form

This also improves default size of the dialog

SjurdurS committed:
Adjust default size of Shape Selection Form

SjurdurS committed:
Formatting

SjurdurS committed:
Remember form size

SjurdurS committed:
Try to make sure Shape Selection Form is visible on open

SjurdurS committed:
Scale vertical scroll bar based on calculated dpi scaling factor

This is a quick and dirty hack, and might not work in all cases

SjurdurS committed:
Remove Across level restriction when loading a shape

SjurdurS committed:
Cleanup comments

SjurdurS committed:
Use SaveFileDialog used for saving levels, instead of the one used for images

SjurdurS committed:
Remove unnecessary path extension modification

SjurdurS committed:
Update comment with correct information

SjurdurS committed:
Remove no longer necessary if statement

SjurdurS committed:
Prevent shapes from being loaded immediately after selecting a shape in the Shape Selection Form

SjurdurS committed:
Fix incorrect highlighting of selected shape when scrolling through shape groups

SjurdurS committed:
Refactor for loop

SjurdurS committed:
Remove no longer needed private member variable

SjurdurS committed:
Discard unused parameter

SjurdurS committed:
Refactor to use Level object instead of individual polygon, objects and grapich elements lists

SjurdurS committed:
Remove no longer needed CalculateBoundingBox method

SjurdurS commented:
I think I've fixed those scaling issues now. I had big problems with margins/padding etc, so I just ended up removing them altogether.

Size should be remembered now as well (per session only).

I also did some refactoring to cleanup the code a bit.

Smibu commented:
Nice :) it looks just fine without margins or paddings.

Some more stuff to tweak:

* Rename "Create Custom Shape" -> "Save as shape..."
* If the shape dir does not exist, it should create it before prompting for filename.
* If the shape dir is empty, it should create a dir "Uncategorized" and use that as the default dir. Because if you save the shape to the root dir, it's ignored. This way the user hopefully learns how it works.
* Default filename should be something like "Type shape title here.lev" so that it's more obvious that the filename will be used as the title.
* If there are no shapes, shape window could say something like "No shapes found. Select + right-click in editor to save selection as a new shape.". Or can also be just a message box if it's easier - no need to open the window because you can't do anything there anyway without shapes.
* Shape tool should have mnemonic (1-letter shortcut that is underlined when pressing Alt) like the other tools. Letter `h` seems fine for this.

Some minor things but would be nice to get fixed if not too hard:

* Shape ordering is not optimal, for example it shows "Eyes 1", "Eyes 11", "Eyes 12", etc. instead of "Eyes 1", "Eyes 2", etc.
* If squeezing the window horizontally, zoom fill doesn't quite work: ![image](https://github.com/user-attachments/assets/2f41b158-1abb-4f99-a2a0-59060c188329)

SjurdurS committed:
Add mnemonic 'h' to shape tool

SjurdurS committed:
Remove unnecessary initialization code

This was just used for testing when trying the new GLControl

SjurdurS committed:
Change the string for saving a shape

SjurdurS committed:
Change default suggested file name for new shapes

SjurdurS committed:
Create shapes directory if it doesn't exist

SjurdurS committed:
Create a default "Uncategorized" directory if shapes directory is empty

SjurdurS committed:
Show a warning when using Shape tool if no shapes are found

SjurdurS committed:
Use natural sort for shape names

SjurdurS committed:
Remove no longer used calculation of padding/margins

SjurdurS committed:
Update zoom fill when resizing LevelControl

SjurdurS committed:
Fix compiler warnings

Smibu commented:
Thanks, just one issue I noticed - when saving shape, the default dir is still the root dir, rather than a subdir.

So IMO, it should never offer the root dir by default, but some subdir (either last used or Uncategorized as fallback).

And maybe additionally, if you try to save the shape to the root dir (or even somewhere outside of it), it should give some error and prompt again. We should make it hard for the user to use it incorrectly.

SjurdurS committed:
Fix incorrect calculation of scroll bar height

SjurdurS committed:
Make sure previously selected shape is visible on open

SjurdurS committed:
Remove unused code

SjurdurS committed:
Add a new item to the shapes dropdown list to view all shapes in one view

SjurdurS committed:
Ensure shapes root folder is relative to executable directory

SjurdurS committed:
Formatting

SjurdurS committed:
Prevent user from saving shapes outside of or at the root level of the sle_shapes directory

SjurdurS committed:
Remember last used shapes directory when saving shapes, or default to Uncategorized if the root directory was just created

SjurdurS committed:
Formatting

SjurdurS committed:
Reset variable that holds the last used shapes directory if the directory no longer exists

SjurdurS commented:
> Thanks, just one issue I noticed - when saving shape, the default dir is still the root dir, rather than a subdir.
> 
> So IMO, it should never offer the root dir by default, but some subdir (either last used or Uncategorized as fallback).
> 
> And maybe additionally, if you try to save the shape to the root dir (or even somewhere outside of it), it should give some error and prompt again. We should make it hard for the user to use it incorrectly.

It should now remember the last used sub-directory, or default to Uncategorized when this folder was just created.
It will still initially suggest the root directory. I hope it makes it easier for the user to see where they can save the shapes, and not just default them to always save to the "Uncategorized" folder.

Also, I added a new "All Shapes" item to the Shapes categories dropdown list, to view all shapes in all sub-folders in one view.

SjurdurS committed:
Temporarily disable redraw of table layout to prevent the custom shape controls appearing one by one in some cases

SjurdurS committed:
Fixed an issue where the wrong shape is selected when closing Shape Selection Form with the OK button

Smibu commented:
That sounds good. The tool works pretty well now. Functionality-wise, this is starting to feel complete.

As the next step (if you didn't already), you could do some sort of quick self-review for the code and see if there's anything super obvious left to refactor/fix. After fixes, update the PR branch and I'll review hopefully soon-ish.

SjurdurS committed:
Code cleanup

SjurdurS committed:
Code cleanup

SjurdurS committed:
Add clarifying comment

SjurdurS committed:
Move common code outside of if statement

SjurdurS committed:
Remove unused using statements

SjurdurS committed:
Remove unused using statements

SjurdurS committed:
Remove unreferenced local variable

SjurdurS committed:
Make debugging function static

SjurdurS committed:
Make fields readonly

SjurdurS committed:
Convert to file-scoped namespace to be consistent with the rest of the code

SjurdurS committed:
Remove unused using statements

SjurdurS committed:
Remove no longer used fields

SjurdurS committed:
Reformatting

SjurdurS committed:
Sort using statements

SjurdurS committed:
Make fields readonly

SjurdurS committed:
Make field readonly

SjurdurS committed:
Code cleanup

SjurdurS committed:
Fixed a problem where the form might appear outside of the screen borders after having resized the form

SjurdurS committed:
Refactoring and code cleanup

SjurdurS committed:
Add debugging statement

SjurdurS committed:
Comment out dispose call that causes a crash in some rare cases

SjurdurS committed:
Remove redundant call to UpdateBounds

SjurdurS committed:
Code cleanup and refactoring

SjurdurS commented:
I’ll try to do some cleaning but I feel it’s a never ending process as there is always something you can do better. :)

I also noticed there is an issue with the Anchoring feature when moving a shape around before placing, the anchoring doesn’t take Size changes into consideration. So if you make the shape smaller or bigger, the anchoring is “off”.

Grass rendering also doesn’t work properly during shape rendering (when you move the shape around before placing it. Not sure how to fix this one.

SjurdurS committed:
Fix anchor offset issue when shape is scaled larger or smaller

SjurdurS committed:
Refactor saving of shapes to use its own SaveFileDialog instance

SjurdurS committed:
Rename variable to reflect stored file path instead of name

Previously, the variable stored the shape name, but it is more convenient to use the file path.

SjurdurS committed:
Regrouping of fields

SjurdurS committed:
Regrouping of fields

SjurdurS committed:
Regrouping of fields

SjurdurS committed:
Merge branch 'refactor_to_virtual_mode_panel' into define_custom_shapes

SjurdurS commented:
I've now done a bit of refactoring/cleaning, and fixed that Anchoring issue.

The pull request branch has been updated with the latest changes :)

Smibu commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,211 @@
+    private readonly RenderingSettings _renderingSettings;
+    private readonly SceneSettings _sceneSettings;
+    private ZoomController? _zoomController;
+    private bool _isFirstRender = true;
```
Can we avoid this field at all?

SjurdurS replied:
When I introduced I did it because the _renderer.UpdateSettings(...) call was slow, and I only wanted it to happen later (after the initialization process). But the performance seems to be OK without this field, so I'll remove it and simplify the code.

SjurdurS committed:
Refactor to fewer nullable fields

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,211 @@
+    private bool _isFirstRender = true;
+    public bool DisableRendering { get; set; } = true;
+
+    private DispatcherTimer? _resizeTimer;
```
On my computer it is quite much slower without the timer. But we can remove it for now to simplify the code. It can always be re-introduced (or some other fix) if anyone complains about it.

SjurdurS committed:
Remove debouncing of resize events

SjurdurS committed:
Revert access modifier changes and move assignments to a new constructor overload instead

SjurdurS committed:
Refactor to no longer update rendering context at first render

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,211 @@
+    private void LevelControl_Load(object? sender, EventArgs e)
+    {
+        // Initialization code for OpenGL
+        _renderer = new ElmaRenderer(this, _renderingSettings);
+        _renderer.OpenGlLgr = _originalElmaRenderer.OpenGlLgr; // Slightly faster with these lines. There is however likely a memory leak here
+        _renderer._lgrCache = _originalElmaRenderer._lgrCache; // Slightly faster with these lines. There is however likely a memory leak here
```
I've refactored this to use a constructor overload. I guess that is what you meant with your example

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeSelectionForm.cs:
```diff
@@ -0,0 +1,651 @@
+    // List of currently active shapes
+    private List<(string Name, string FilePath)> _shapes;
+
+    public ShapeSelectionForm(GLControl sharedContext, ElmaRenderer elmaRenderer, string? selectedShapeFilePath = null, double scalingFactor = 0.0, double rotationAngle = 0.0, ShapeMirrorOption mirrorOption = ShapeMirrorOption.None)
```
You are right :D Little brain fart from me there. I don't know why I thought I still needed those. I must have thought they would need to be passed back and forth between CustomShapeTool and the form, but of course the values are already stored in CustomShapeTool.. Good to have another set of eyes look at the code :)

SjurdurS committed:
Remove no longer needed fields

SjurdurS committed:
Refactor to non-nullable fields

SjurdurS committed:
Revert package upgrades

SjurdurS commented on OpenTK.WinForms/OpenTK.WinForms.csproj:
```diff
@@ -26,8 +26,9 @@
   <ItemGroup>
-    <PackageReference Include="OpenTK.Graphics" Version="4.8.2" />
-    <PackageReference Include="OpenTK.Windowing.Desktop" Version="4.8.2" />
+    <PackageReference Include="OpenTK.GLControl" Version="4.0.2" />
+    <PackageReference Include="OpenTK.Graphics" Version="4.9.3" />
+    <PackageReference Include="OpenTK.Windowing.Desktop" Version="4.9.3" />
```
I thought we updated to make use of SharedContext. Anyways, it works fine after reverting these changes, so I'll just revert it.

SjurdurS committed:
Refactor ShapeDataDto to SleShape and move deserialization logic to this class

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeDataDto.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Elmanager.LevelEditor.Shapes;
+
+internal class ShapeDataDto(Level level)
```
Something like this?

b2fab90b5e9dcc5a9524783720c57965a70ec9c1

SjurdurS committed:
Center ShapeSelectionForm to parent

Remove manual calculation of form position and just center the form to the parent form instead

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeSelectionForm.cs:
```diff
@@ -0,0 +1,651 @@
+        }
+    }
+
+    private void SetInitialLocation()
```
Fixed in d058e9250701d45435231b307d0340ea01cd9c5e

SjurdurS committed:
Refactor to remove unnecessary manual handling of form cancel

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeSelectionForm.cs:
```diff
@@ -0,0 +1,651 @@
+        this.Close();
+    }
+
+    protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
```
Fixed in 0de91f7bf53990ccaf57d14be64a49f5ad8e8b9e

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeSelectionForm.cs:
```diff
@@ -0,0 +1,651 @@
+        return CompareNatural(strA, strB, CultureInfo.CurrentCulture, CompareOptions.IgnoreCase);
+    }
+
+    public static int CompareNatural(string strA, string strB, CultureInfo culture, CompareOptions options)
```
I didn't want to install a new package just for this, but if you don't mind then why not. Although that package doesn't handling leading zeros. This one seems better:
https://github.com/tompazourek/NaturalSort.Extension

SjurdurS replied:
Fixed in 5ace6523d06be5df373ba54d0d6abb40f184b662

SjurdurS committed:
Add nuget package for handling NaturalSort

SjurdurS committed:
Move Shape saving to CustomShapeTool

SjurdurS committed:
Remove no longer needed selection state handling

This was used to take bmp snapshot of the current selection without showing selection markers in the picture.
Shapes are now saved as Levels so this is no longer a problem.

SjurdurS commented on Elmanager/LevelEditor/LevelEditorForm.cs:
```diff
@@ -2953,4 +2969,149 @@ private void fixSelfIntersectionsMenuItem_Click(object sender, EventArgs e)
+
+    private void createCustomShapeMenuItem_Click(object sender, EventArgs e)
+    {
+        var selectedPolygons = Lev.Polygons.Where(p => p.Vertices.Any(v => v.Mark == VectorMark.Selected)).ToList();
```
Fixed in 12681d1230248fb6e2c122d1136b91bd86275ede

SjurdurS commented on Elmanager/LevelEditor/LevelEditorForm.cs:
```diff
@@ -87,6 +87,7 @@ internal partial class LevelEditorForm : FormMod, IMessageFilter
     private ToolBase.NearestVertexInfo? _grassInfo;
     private TexturizationOptions? _texturizationOpts;
     private readonly LevFileWatcher _levFileWatcher;
+    private string? _lastUsedShapeFolder;
```
Fixed in 12681d1230248fb6e2c122d1136b91bd86275ede

SjurdurS committed:
Fix auto generated using statement

SjurdurS committed:
Cleanup comments

SjurdurS commented on Elmanager/LevelEditor/Shapes/CustomShapeControl.cs:
```diff
@@ -0,0 +1,199 @@
+using System.Windows.Forms;
+using Color = System.Drawing.Color;
+using Pen = System.Drawing.Pen;
+using UserControl = System.Windows.Forms;
```
Fixed in f4cc630ae0e156734d94d0621cd14ba8ccc3dddf

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeSelectionForm.cs:
```diff
@@ -0,0 +1,651 @@
+
+    private void PopulateSubfolderComboBox()
+    {
+        // Clear existing items
```
I removed some comments in this commit:
8e4195d1751c2e2516629ddb4690d18af12de294

SjurdurS committed:
Cleanup comments

SjurdurS committed:
Refactor to non-nullable fields

SjurdurS commented:
I've gone through all of your proposed changes now. Let me know if I should change anything else.

Smibu commented on Elmanager/LevelEditor/Shapes/CustomShapeControl.cs:
```diff
@@ -0,0 +1,189 @@
+        this.DoubleClick += CustomShapeControl_DoubleClick;
+    }
+
+    public string ShapeFullPath { get; set; } = String.Empty;
```
I'd rather make this nullable instead of using magic invalid value:

```suggestion
    public string? ShapeFullPath { get; set; };
```

SjurdurS replied:
Fixed in c50fec9d24f45188b8d89ecb9e015aa826370299

Smibu commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,458 @@
+        if (!hasLevFiles)
+        {
+            UiUtils.ShowError("No .lev files found in any subdirectory of 'sle_shapes'.\nSelect + right-click in editor to save selection as a new shape.",
+                "No shapes found", MessageBoxIcon.Exclamation);
```
```suggestion
                "No shapes found", MessageBoxIcon.Information);
```

Smibu commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,458 @@
+            return;
+        }
+
+        // Open Shape Selection Form
```
```suggestion
```

SjurdurS replied:
I know it's an obvious comment, I just sometimes use in-line comments like this as visual cues to separate sections of code. I'll remove it :)

Smibu commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,458 @@
+        }
+
+        // Open Shape Selection Form
+        var shapeSelectionForm = new ShapeSelectionForm(LevEditor.EditorControl, LevEditor.Renderer, _selectedShapeFilePath);
+        shapeSelectionForm.ShapeDataLoaded += ShapeSelectionForm_ShapeDataLoaded;
+        shapeSelectionForm.ShowDialog(LevEditor);
```
Instead of having a ShapeDataLoaded event like this, could we use the same pattern as in e.g. SvgImportOptionsForm? So just a static method that opens the form and returns the data (or null if canceled).

SjurdurS replied:
Not sure I got exactly how you wanted it, but something like this c91d875a65a9cc8b44356985571fd973804c7754 ?

SjurdurS replied:
Additional changes made here: e0f79a3cf64ba4808b38015edf3523f38a6b0d9b 

Removed an additional ShapeDataLoaded event and deserialization of shapes is now purely handled by the SleShape class

Smibu reviewed (commented):
Thanks! I added some more comments.

SjurdurS committed:
Update Elmanager/LevelEditor/Shapes/LevelControl.cs

Remove function comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,173 @@
+
+        if (!IsHandleCreated)
+        {
+            SharedContext = sharedContext; // Set shared context before initialization
```
From what I can find the SharedContext must only be set before the handle has been created. I don't know C#/Winforms well enough to know whether this is guaranteed to be the case at the call of this line.

```
        /// <summary>
        /// Gets or sets the <see cref="GLControl"/> used to share OpenGL resources.
        /// This value cannot be changed after the control has been initialized (before <see cref="OnHandleCreated(EventArgs)"/> is triggered).
        /// </summary>
        [Category("OpenGL")]
        public GLControl? SharedContext
        {
            get => _sharedContext;
            set
            {
                if (_nativeWindow == null)
                {
                    _sharedContext = value;
                }
                else
                {
                    throw new InvalidOperationException("Can't set OpenGL settings when the control is initialized.");
                }
            }
        }
        private GLControl? _sharedContext;

```

Smibu replied:
I think [this](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.control.createcontrol?view=windowsdesktop-9.0#system-windows-forms-control-createcontrol) answers the question:

> simply calling a control's constructor does not create the Handle.

So let's remove the if :)

SjurdurS committed:
Remove debugging statement

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Remove redundant Gl.Clear call

SjurdurS committed:
Remove no longer used GL debugging function

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit);
+        CheckGLError("GL.Clear");
```
Fixed in dbc5bce87fae22a2346bede059de323d77a5357c

SjurdurS committed:
Remove comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Remove comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Change message box icon

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Use pattern matching instead of HasValue + Value

SjurdurS commented on Elmanager/LevelEditor/Shapes/ShapeSelectionForm.cs:
```diff
@@ -0,0 +1,411 @@
+        if (_lastSize.HasValue)
+        {
+            this.Size = _lastSize.Value;
+        }
```
Like this 529b1535007e2631cd6f2a936b13d53d840ef504  ?

Seems very unreadable to me, but I'm also not used to C#.

SjurdurS committed:
Simplify ElmaRenderer initialization with lambda in LevelControl

SjurdurS committed:
Discard unused lambda parameters

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,173 @@
+
+        _renderer = elmaRenderer;
+
+        this.Load += LevelControl_Load;
```
Fixed applied in 48b9ee991ae7943c700a0c3beb225431bac6026a

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,173 @@
+        _renderer.DrawScene(_level, _camera, _sceneSettings, _renderingSettings);
+    }
+
+    private static void CheckGLError(string location)
```
Removed in 5b742009945c36664e1883eb6fcbe7f3444c3973

SjurdurS committed:
Refactor viewport resetting to own function

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,173 @@
+        if (resetViewport)
+        {
+            GL.Viewport(0, 0, Width, Height);
+            _zoomController.ZoomFill(_renderingSettings);
```
Fixed in ccc2eb3ed49314ab5c7e108c930434917fe03026

SjurdurS committed:
Code cleanup

Remove unused parameters and change access modifier

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Remove unnecessary Dispose override

The OpenGL resources are shared, so it makes sense that we can't destroy them as usual.

SjurdurS commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,173 @@
+            //_renderer.Dispose();
+        }
+        base.Dispose(disposing);
+    }
```
Good point. I've removed it.

SjurdurS committed:
Fix formatting of Custom Shape Tool help string

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Remove comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Remove comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Refactor to use static method to open form

SjurdurS committed:
Change to nullable field

SjurdurS committed:
Remove no longer used ShapeDataLoaded event

Also cleanup no longer used code when form is closed with the OK button

SjurdurS committed:
Refactor to use SleShape instead of Level where applicable

SjurdurS commented on Elmanager/LevelEditor/Shapes/CustomShapeControl.cs:
```diff
@@ -0,0 +1,189 @@
+        SetLevel(level);
+    }
+
+    internal void SetLevel(Level level)
```
0424374ca69d7c662435510923201ba2e028204e

SjurdurS committed:
Refactor shape data fields and remove level fields

SjurdurS commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,458 @@
+
+    // Level
+    private Level _level;
+    private Level _originalLevel;
```
A bit of a change, but it does seem a bit cleaner now. Could maybe be even more clean but I'm running out of time for this programming session :)

fdb64b2ea30f092868351344a464d8987ce814bb

SjurdurS committed:
Rename function to save shapes

SjurdurS commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,458 @@
+        lev.Save(filePath);
+    }
+
+    public void CreateCustomShapeMenuItem_Click(object sender, EventArgs e)
```
b7de2dfc67cafa0c220a1c736bd6bc7073e11f0e

SjurdurS commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,458 @@
+    private void SaveShape(Level lev, string filePath)
+    {
+        lev.Save(filePath);
+    }
```
b7de2dfc67cafa0c220a1c736bd6bc7073e11f0e

SjurdurS commented:
> Thanks! I added some more comments.

I think I've gone through / replied to all of the comments you made now.

SjurdurS committed:
Remove use of this keyword

SjurdurS committed:
Refactor inline color definitions with the corresponding fields

SjurdurS committed:
Try to generalize column/row setup and calculations to make it easier to adjust if necessary

SjurdurS committed:
Change back to 4x3 setup

Smibu commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,440 @@
+
+    public void ExtraRendering()
+    {
+        if (_shapeSelection == null)
+        {
+            return;
+        }
+
+        var settings = Global.AppSettings.LevelEditor.RenderingSettings;
+        foreach (var polygon in _shapeSelection.Shape.Obj.Level.Polygons)
+        {
+            if (polygon.IsGrass)
+            {
+                if (settings.ShowGrassEdges)
+                {
+                    var color = polygon.SlopeInfo?.HasError ?? false ? Color.Red : settings.GrassEdgeColor;
+                    Renderer.DrawGrassPolygon(polygon, color, settings.ShowInactiveGrassEdges, settings);
+                }
+            }
+            else
+            {
+                if (settings.ShowGroundEdges)
+                {
+                    Renderer.DrawPolygon(polygon, settings.GroundEdgeColor);
+                }
+            }
+        }
```
ExtraRendering shouldn't be needed because we have transient elements already and we don't have to render anything extra

```suggestion
```

SjurdurS replied:
Grass doesn't show at all without this. Before it would show at least the grass vertices (but not grass pictures).

Before (With extra rendering):
https://github.com/user-attachments/assets/f70cf694-4376-4c6f-b7e2-f36a1c75ca98



After (Without extra rendering:
https://github.com/user-attachments/assets/0c7a9959-d4da-4a93-9492-8dbaea6ce759

Smibu replied:
Found the problem:

* you need to call UpdateDecompositionOrGrassSlopeInfo instead of just UpdateDecomposition
* in ElmaRenderer at line 455 it draws grass edges but it doesn't account for TransientElements so you have to concat them to the main list, see examples elsewhere in that file

SjurdurS replied:
> Found the problem:
> 
>     * you need to call UpdateDecompositionOrGrassSlopeInfo instead of just UpdateDecomposition
> 
>     * in ElmaRenderer at line 455 it draws grass edges but it doesn't account for TransientElements so you have to concat them to the main list, see examples elsewhere in that file

I've now updated ApplyTransformations to call UpdateDecompositionOrGrassSlopeInfo instead of UpdateDecomposition.

But I'm not sure I understand what you mean about the concatenating part.

SjurdurS replied:
There is also another issue which I haven't looked into yet. Images/Textures in shapes are rendered as wire-frames while moving the shape around. Also, the images do not even get updated to real images when the shape is placed. If you have any tips how to fix this I'm all ears, otherwise I'll look into it later.

SjurdurS replied:
I've tried to load images now, but there is another issues with LGR. If you use default lgr then it looks fine, but if you use any other LGR, then when opening the Shape Selection Form all images go blank. I think it's because the loaded shape defaults to the "default" lgr, and it affects the elmarenderer.

I made a quick (but not the most elegant) fix for this here:  95fef2accaf4e420eb3d2fb4bcda7dc47813f629 

Now, if you switch LGR while a shape is loaded, then images go blank. I think a fix for this would be to notify CustomShapeTool when the global LGR is changed.

Smibu replied:
> But I'm not sure I understand what you mean about the concatenating part.

For example ElmaRenderer line 256:

foreach (var k in lev.Polygons.Concat(sceneSettings.TransientElements.Polygons))

This concat is missing in line 455 foreach loop.

I'll check the image problem sometime.

SjurdurS replied:
> > But I'm not sure I understand what you mean about the concatenating part.
> 
> For example ElmaRenderer line 256:
> 
> foreach (var k in lev.Polygons.Concat(sceneSettings.TransientElements.Polygons))
> 
> This concat is missing in line 455 foreach loop.
> 
> I'll check the image problem sometime.

I understand, it seems to work now: 17e016a394421e187e75212bd3f19cd2d5fc1efe

SjurdurS replied:
The image issue after changing lgr also happens with the PictureTool so its nothing new.

Smibu replied:
Ah yea, you're right. No need to fix in this PR.

Smibu commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,440 @@
+        }
+
+        Level level = _shapeSelection.Shape.Obj.Level;
+        return new TransientElements(new List<Polygon>(level.Polygons), new List<LevObject>(level.Objects),
+            new List<GraphicElement>(level.GraphicElements));
```
Do we need to clone (`new List<...>`) here? TransientElements is used just for rendering so I think it should be fine to skip cloning

SjurdurS replied:
To be honest I never really checked. I just wrote it like that based on some of your existing code.
Like in TransientElements.cs

SjurdurS replied:
I've removed the cloning.

Smibu commented on Elmanager/Rendering/ElmaRenderer.cs:
```diff
@@ -49,6 +49,15 @@ internal ElmaRenderer(GLControl renderingTarget, RenderingSettings settings)
     }
+    internal ElmaRenderer(GLControl renderingTarget, RenderingSettings settings, ElmaRenderer otherElmaRenderer)
+    {
+        _gfxContext = renderingTarget.Context;
+        InitializeOpengl(disableFrameBuffer: settings.DisableFrameBuffer);
```
Could we call the other constructor here because 2 first lines are the same?

SjurdurS replied:
Good point. Done.

Smibu commented on Elmanager/LevelEditor/Tools/CustomShapeTool.cs:
```diff
@@ -0,0 +1,440 @@
+
+    internal CustomShapeTool(LevelEditorForm editorForm) : base(editorForm)
+    {
+        _shapeSelection = null;
```
Null is default so I'd just drop this

```suggestion
```

SjurdurS replied:
Removed.

Smibu commented on Elmanager/LevelEditor/Shapes/SleShape.cs:
```diff
@@ -0,0 +1,73 @@
+
+        // Normalize positions
+        List<Polygon> levelPolygons = level.Polygons;
+        List<LevObject> levObjects = level.Objects; // No need to filter start object since it is already removed
```
```suggestion
        List<LevObject> levObjects = level.Objects;
```

Smibu reviewed (commented):
Thanks for the updates, it's getting closer to finish line :) I added a few more comments here.

Found a bug (imo): if I place a shape, reopen form and cancel, it forgets the last selected shape. I think this used to work so maybe broke in refactoring.

SjurdurS commented:
> Thanks for the updates, it's getting closer to finish line :) I added a few more comments here.
> 
> Found a bug (imo): if I place a shape, reopen form and cancel, it forgets the last selected shape. I think this used to work so maybe broke in refactoring.

Yeah I know, it happened after refactoring to storing the state in the _shapeSelection object. I left it for now but I can see if I can fix it.

SjurdurS committed:
Use file-scoped namespace

SjurdurS committed:
Remove extra rendering

ExtraRendering shouldn't be needed because we have transient elements already and we don't have to render anything extra

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Refactor constructor to call other constructor for shared initialization

SjurdurS committed:
Remove redundant null initialization

SjurdurS committed:
Remove comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS committed:
Remove unnecessary if-statement

The handle should never be able to be created at this point anyways

SjurdurS committed:
Remove unnecessary cloning of objects

SjurdurS committed:
Revert "Remove extra rendering
"

This reverts commit a5146ac9b0d4f4a226b0fab7f33b2f00a73f7e3a.

SjurdurS committed:
Remove unnecessary non-grass polygon rendering in extra rendering

SjurdurS committed:
Fixed an issue where the currently selected shape would reset to no selection after closing Shape Selection From with the Cancel button

SjurdurS commented:
> > Thanks for the updates, it's getting closer to finish line :) I added a few more comments here.
> > Found a bug (imo): if I place a shape, reopen form and cancel, it forgets the last selected shape. I think this used to work so maybe broke in refactoring.
> 
> Yeah I know, it happened after refactoring to storing the state in the _shapeSelection object. I left it for now but I can see if I can fix it.

The bug has been fixed here: 3fe32b0e07552f7c94c1a147b393b961ebd074d7

SjurdurS committed:
Remove and sort usings

SjurdurS committed:
Whitespace formatting

SjurdurS committed:
Fix an issue rendering grass

SjurdurS committed:
Use current level bounds and not shape bounds for updating grass slope info

SjurdurS committed:
Update images when shape is loaded and placed

SjurdurS committed:
Fix an issue where images and textures incorrectly updated to using wrong LGR when opening Shape Selection Form

SjurdurS committed:
Fix grass rendering and remove no longer needed extra rendering

SjurdurS committed:
Reset anchoring when resetting shape transformations

Smibu commented on Elmanager/LevelEditor/Shapes/LevelControl.cs:
```diff
@@ -0,0 +1,145 @@
+            UpdateRenderingContext();
+        };
+
+        SharedContext = sharedContext; // Set shared context before initialization
```
```suggestion
        SharedContext = sharedContext;
```

Smibu reviewed (commented):
Thanks! This seems almost ready for merging. Can you update PR description to reflect current state? I will squash merge this, so PR title and description will be used as the commit message.

SjurdurS committed:
Remove comment

Co-authored-by: Mika Lehtinen <Smibu@users.noreply.github.com>

SjurdurS commented:
> Thanks! This seems almost ready for merging. Can you update PR description to reflect current state? I will squash merge this, so PR title and description will be used as the commit message.

I've updated the pull request description (the first message in this pull request). Let me know if I should change anything :)

Smibu commented:
Big thanks and congrats for getting this over the finish line. :)

I will create a build later today and announce in discord.

SjurdurS commented:
Thank you for all your help! Very much appreciated the time you put in to helping me get this finished :)
